doc/neps/nep-0055-string_dtype.rst

```diff
@@ -0,0 +1,855 @@
+left alone due to backward compatibility concerns, although the unicode DType
+became the default DType for ``str`` data and the old ``string`` DType was
+renamed the ``bytes_`` DType. This change left NumPy with the sub-optimal
+situation of shipping a data type originally intended for null-terminated
+bytestrings as the data type for *all* python ``bytes`` data, and a default
+string type with an in-memory representation that consumes four times as much
+memory as needed for ASCII or mostly-ASCII data.
+
+Problems with Fixed-Width Strings
+*********************************
+
+Both existing string DTypes represent fixed-width sequences, allowing storage of
+the string data in the array buffer. This avoids adding out-of-band storage to
+NumPy, however, it makes for an awkward user interface. In particular, the
+maximum string size must be inferred by NumPy or estimated by the user before
+loading the data into a NumPy array or selecting an output DType for string
+operations. In the worst case, this requires an expensive pass over the full
+dataset to calculate the maximum length of an array element. It also wastes
+memory when array elements have varying lengths. Pathological cases where an
+array stores many short strings and a few very long strings are particularly bad
+for wasting memory.
+
+Downstream usage of string data in NumPy arrays has proven out the need for a
+variable-width string data type. In practice, most downstream users employ
+``object`` arrays for this purpose. In particular, ``pandas`` has explicitly
+deprecated support for NumPy fixed-width strings, coerces NumPy fixed-width
+string arrays to ``object`` arrays, and in the future may switch to only
+supporting string data via ``PyArrow``, which has native support for UTF-8
+encoded variable-width string arrays [1]_. This is unfortunate, since ``object``
+arrays have no type guarantees, necessitating expensive sanitization passes and
+operations using object arrays cannot release the GIL.
+
+Previous Discussions
+--------------------
+
+The project last discussed this topic in depth in 2017, when Julian Taylor
+proposed a fixed-width text data type parameterized by an encoding [2]_. This
+started a wide-ranging discussion about pain points for working with string data
+in NumPy and possible ways forward.
+
+In the end, the discussion identified two use-cases that the current support for
+strings does a poor job of handling:
+
+* Loading or memory-mapping scientific datasets with unknown encoding,
+* Working with string data in a manner that allows transparent conversion
+  between NumPy arrays and Python strings, including support for missing
+  strings.
+
+As a result of this discussion, improving support for string data was added to
+the NumPy project roadmap [3]_, with an explicit call-out to add a DType better
+suited to memory-mapping bytes with any or no encoding, and a variable-width
+string DType that supports missing data to replace usages of object string
+arrays.
```

Comment:
@ngoldbaum, I'm reading some of the references to have better context for a review of the NEP.  I don't see any comment in the roadmap about strings supporting missing data, and I didn't see it in the 2017 mega-thread.  Was missing data for strings specifically discussed somewhere else?  Or are you referring to the items about `MaskedArray`s in the Maintenance section of the roadmap, where a bullet point is "dtypes that support missing values".  I guess the implicit suggestion is to use missing values *instead of* masked arrays?

Reply:
I'm not sure offhand if missing data support is mentioned in the discussions you're looking at. The morivation for adding it wasn't previous discussions so much as writing a functioning version of Pandas that supports the new dtype. It may be possible to have a version of Pandas that handles missing data without explicit support in the dtype, but I don't think it's possible to do without substantial performance implications.

I don't think I ever imply in the text of the NEP that previous discussions asked for a string dtype with missing data support. If you'd like I can add text explicitly mentioning that.

Reply:
The text in the NEP seems to suggest that adding the capability for missing data was discussed in the April 2017 discussion, but I didn't see any mention of missing data in there.

To be clear, I'm definitely in favor of adding some form of missing data or "not a string" capability.  But for the purpose of reviewing the NEP, I'm interested in what has already been discussed about this feature, to ensure that any specific needs or use-cases that have already been mentioned are addressed.

Reply:
It did come up, e.g. [here](https://mail.python.org/pipermail/numpy-discussion/2017-April/076681.html). I'd also argue that the requirement that it needs to replace usages of object arrays implicitly requires supporting missing data, since object arrays get that for free (albeit with no type checking).

Reply:
Ah, so this line from @rkern 

> * Maybe add a dtype similar to `object_` that only permits `unicode/str`
(2.x/3.x) strings (and maybe None to represent missing data a la pandas).

Thanks for finding that.  That counts as "discussion" I guess, but I thought I would find a bit more.

Reply:
There may be others, that was just a quick search by me on my phone...

Reply:
When I say something, it Has Been Discussed. :wink: 

More seriously, consider it a proxy for all of the design discussion that went behind Pandas' choice to use object arrays for strings and then for Arrow to support nulls in `StringArray`s. They have a real use case, and improving their situation should (I think) be a core motivation of this NEP.